### PR TITLE
Add support for BitBucket Webhooks

### PR DESCRIPTION
**Pull request by Forest-Troll**
##### SUMMARY
Restore commit https://github.com/ansible/awx/commit/885841caeaea0f7c86d39d6e7a66b960124289c2  with some additional fix for integrate with Bitbucket Server.
related #5209

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - API

##### AWX VERSION

```
awx: 15.0.0
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: awx/main/migrations/0092_v360_webhook_mixin.py [hunk 1]
```diff
@@ -39,7 +39,7 @@ class Migration(migrations.Migration):
         migrations.AddField(
             model_name='workflowjobtemplate',
             name='webhook_service',
-            field=models.CharField(blank=True, choices=[('github', 'GitHub'), ('gitlab', 'GitLab')], help_text='Service that webhook requests will be accepted from', max_length=16),
+            field=models.CharField(blank=True, choices=[('github', 'GitHub'), ('gitlab', 'GitLab'), ('bitbucket', 'Bitbucket')], help_text='Service that webhook requests will be accepted from', max_length=16),
```
Comment (wenottingham) [inline awx/main/migrations/0092_v360_webhook_mixin.py:42]: I do not believe you can rewrite past migrations in this way.

Comment (Forest-Troll) [reply awx/main/migrations/0092_v360_webhook_mixin.py:42]: But hooks with BitbucketServers was worked.
But callback status to BitbucketServer not working yet...

I changed cipher for has validate to sha256, and added empty method for get_event_status_api, nothing more.

Oh, I think i know what you mean :-)

Comment (ryanpetrello) [reply awx/main/migrations/0092_v360_webhook_mixin.py:42]: No, this needs to be a new migration.  You should be able to generate one using `awx-manage makemigrations`.

**Comment by ryanpetrello** [comment]
@jbradberry I remember we looked at this initially, but eventually yanked support for some reason.  Do either of you remember why, and does this PR address it (or has some needed support changed on the Bitbucket end, somehow?)

**Comment by jbradberry** [comment]
> @jbradberry I remember we looked at this initially, but eventually yanked support for some reason. Do either of you remember why, and does this PR address it (or has some needed support changed on the Bitbucket end, somehow?)

Yes, we needed to obtain a license for Bitbucket Server and it didn't look like we were going to get it in time.  The free website version of Bitbucket does not support signed webhooks, so support for that should be excluded, if that can be done easily.  It also wasn't clear without trying it out whether the machinery that I built that worked for Github and Gitlab would work for Bitbucket without serious elbow grease.

I'm pretty sure that this PR is just doing a `git revert ...` on the commit where I yanked out the Bitbucket code and doesn't solve the problems of 1) knowing whether the code I wrote blindly would even work correctly, and 2) migrating data that has already landed.

**Comment by ryanpetrello** [comment]
> The free website version of Bitbucket does not support signed webhooks

Yes, now I remember.  For me, if there's not some _obvious_ way to detect free vs paid and restrict access on the AWX side, then I don't think we should merge this, because I expect the vast majority of users will end up trying to use it with the free version.

*Maybe* we could interpret free vs paid as "if it doesn't _have_ a signature, it is forbidden".  That's a pretty lousy user experience, though, and I'm guessing it'll confuse users (and likely prompt users to ask *us* for an "I don't care about security" button).

Honestly, I'm a little surprised and disappointed that Bitbucket is making people pay for secure defaults (clearly it works if it's in the enterprise product, and they've chosen not to provide it in the free web service), but c’est la vie, I guess.

So to summarize, I'd say we're back at the same place we were a year ago when we decided not to support Bitbucket.

**Review by ryanpetrello — CHANGES_REQUESTED** [review]
Given my comments on this PR, I vote 👎 until Bitbucket can properly support signed web hooks in its free product.

**Comment by opfpqgoon** [comment]
Given that trial licenses of Atlassian products can be requested pretty easily (https://confluence.atlassian.com/bitbucketserver/install-a-bitbucket-server-trial-867192384.html), I don't fully understand why we need to wait on bitbucket cloud to support signed Webhooks.  Given that Atlassian is completely revamping their license model in February, there are quite a few enterprise teams who will be moving from Server licensing to Datacenter licensing.

Why could we not enable bitbucket on-prem (server/datacenter) Webhooks with clear documentation & UI tooltips explaining that this does *not* work with bitbucket cloud?

Having bitbucket Webhooks integration with AWX would massively simplify CI/CD workflows as there's no longer a need for an intermediary service to receive Webhooks and trigger tower-cli commands.

I'd be happy to help test anything related to bitbucket on-prem webhook integration.

**Comment by tech2734** [comment]
I am running into more and more clients that are using Bitbucket and have a pretty clear need for this functionality.   Please reconsider supporting it.  I would help out with testing as well.  thanks.
